compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -969,26 +1110,350 @@ mod test {
+
+        // From loop_2_body:
+        // ```
+        // v9 = mul v0, v1
+        // v11 = mul v9, v0
+        // constrain v11 == u32 12
+        // ```
+        // To loop_1_exit:
+        // ```
+        // v9 = mul v0, v1
+        // v10 = mul v9, v0
+        // constrain v10 == u32 12
+        // ```
+        let expected = "
+      brillig(inline) fn main f0 {
+        entry(v0: u32, v1: u32):
+          v5 = eq v0, u32 5
+          jmp loop_1(u32 0)
+        loop_1(v2: u32):
+          v8 = lt v2, u32 4
+          jmpif v8 then: loop_1_cond, else: loop_1_exit
+        loop_1_cond():
+          jmpif v5 then: loop_1_body, else: loop_1_end
+        loop_1_exit():
+          v9 = mul v0, v1
+          v10 = mul v9, v0
+          constrain v10 == u32 12
+          jmp loop_2(u32 0)
+        loop_1_body():
+          v15 = mul v0, v1
+          constrain v15 == u32 12
+          jmp loop_1_end()
+        loop_1_end():
+          v16 = unchecked_add v2, u32 1
+          jmp loop_1(v16)
+        loop_2(v3: u32):
+          v12 = lt v3, u32 4
+          jmpif v12 then: loop_2_body, else: exit
+        loop_2_body():
+          v14 = unchecked_add v3, u32 1
+          jmp loop_2(v14)
+        exit():
+          return
+      }
+      ";
+
+        assert_normalized_ssa_equals(ssa, expected);
+    }
+
+    #[test]
+    fn do_not_hoist_constrain_in_loop_with_zero_upper_bound() {
+        // This test is the same as `hoist_safe_mul_that_is_non_control_dependent` except
+        // that the upper loop bound is zero
+        let src = "
+        brillig(inline) fn main f0 {
+          entry(v0: u32, v1: u32):
+            jmp loop(u32 0)
+          loop(v2: u32):
+            v3 = lt v2, u32 0
+            jmpif v3 then: loop_body, else: exit
+          loop_body():
+            v6 = mul v0, v1
+            v7 = mul v6, v0
+            constrain v7 == u32 12
+            v10 = unchecked_add v2, u32 1
+            jmp loop(v10)
+          exit():
+            return
+        }
+        ";
+
+        let ssa = Ssa::from_str(src).unwrap();
+
+        let ssa = ssa.loop_invariant_code_motion();
+
+        // We expect the constrain to remain inside of `loop_body`
+        // as the loop is never going to be executed.
+        // If the constrain were to be hoisted out it could potentially
+        // cause the program to fail when it is not meant to fail.
+        let expected = "
+        brillig(inline) fn main f0 {
+          entry(v0: u32, v1: u32):
+            v3 = mul v0, v1
+            v4 = mul v3, v0
+            jmp loop(u32 0)
+          loop(v2: u32):
+            jmpif u1 0 then: loop_body, else: exit
+          loop_body():
+            constrain v4 == u32 12
+            v10 = unchecked_add v2, u32 1
+            jmp loop(v10)
+          exit():
+            return
+        }
+        ";
+
+        assert_normalized_ssa_equals(ssa, expected);
+    }
+
+    #[test]
+    fn do_not_hoist_constrain_in_loop_with_equal_non_zero_loop_bounds() {
```

Comment:
Not sure if we want to keep this test and `do_not_hoist_constrain_in_loop_with_zero_upper_bound`. With this test I wanted to guarantee we were not just checking that the upper bound is zero but that (upper - lower) is not zero. However, it makes the above test feel a bit unnecessary. Let me know if we would prefer both tests or just this one here.

Reply:
I think it's worth keeping both